go/marshal/decode.go

```diff
@@ -0,0 +1,227 @@
+// Copyright 2016 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+package marshal
+
+import (
+	"fmt"
+	"reflect"
+	"runtime"
+	"sync"
+
+	"github.com/attic-labs/noms/go/types"
+)
+
+// Unmarshal converts a Noms value into a Go value. It decodes v and stores the result
+// in the value pointed to by out.
+//
+// Unmarshal uses the inverse of the encodings that Marshal uses with the following additional rules:
+//
+// To unmarshal a Noms struct into a Go struct, Unmarshal matches incoming object
+// fields to the fields used by Marshal (either the struct field name or its tag).
+// Unmarshal will only set exported fields of the struct.
+//
+// If a Noms value is not appropriate for a given target type, or if a Noms number overflows the target type, Unmarshal returns a UnmarshalTypeMismatchError or OverflowError respectively
+//
+func Unmarshal(v types.Value, out interface{}) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			if _, ok := r.(runtime.Error); ok {
+				panic(r)
+			}
+			if s, ok := r.(string); ok {
+				panic(s)
+			}
+			err = r.(error)
+		}
+	}()
+
+	rv := reflect.ValueOf(out)
+	if rv.Kind() != reflect.Ptr || rv.IsNil() {
+		return &InvalidUnmarshalError{reflect.TypeOf(out)}
+	}
+	rv = rv.Elem()
+	d := typeDecoder(rv.Type())
+	d(v, rv)
+	return
+}
+
+// InvalidUnmarshalError describes an invalid argument passed to Unmarshal. (The argument to Unmarshal must be a non-nil pointer.)
+type InvalidUnmarshalError struct {
+	Type reflect.Type
+}
+
+func (e *InvalidUnmarshalError) Error() string {
+	if e.Type == nil {
+		return "Cannot unmarshal into Go nil value"
+	}
+
+	if e.Type.Kind() != reflect.Ptr {
+		return "Cannot unmarshal into Go non pointer of type " + e.Type.String()
+	}
+	return "Cannot unmarshal into Go nil pointer of type " + e.Type.String()
+}
+
+// UnmarshalTypeMismatchError describes a Noms value that was not appropriate for a value of a specific Go type.
+type UnmarshalTypeMismatchError struct {
+	Value   types.Value
+	Type    reflect.Type // type of Go value it could not be assigned to
+	details string
+}
+
+func (e *UnmarshalTypeMismatchError) Error() string {
+	return "Cannot unmarshal " + e.Value.Type().Describe() + " into Go value of type " + e.Type.String() + e.details
+}
+
+// OverflowError describes a Noms value that was not appropriate for a value of a specific Go number type.
+type OverflowError struct {
+	Value types.Number
+	Type  reflect.Type // type of Go value it could not be assigned to
+}
+
+func (e *OverflowError) Error() string {
+	return fmt.Sprintf("Cannot unmarshal %g into Go value of type %s", e.Value, e.Type)
```

Comment:
I wanted 'g' but I can get that through strconv.FormatFloat. I find the Sprintf cleaner though. Maybe I'll change some of the other ones to use that instead.